<center><img src='img/ms_logo.jpeg' width=30% height=30%></center>
<center><h1>Descriptive Statistics</h1></center>

In this lesson, we'll clean and analyze the Titanic data set, and see if descriptive statistics can tell us anything interesting about the data set.  Along the way, we'll gain hands-on experience with industry-standard tools such as the `pandas` library in python. 

#### About the Data Set.  

The Titanic sank after hitting an iceberg the night of May 31st, 1911.  1503 people died--only 705 survived (could have been 706, but some people are door hogs).  

<img src='http://4.bp.blogspot.com/-QZUM4Q23E3c/UJ7WFXdJABI/AAAAAAAAAXk/ityxfFzjpPE/w1200-h630-p-k-no-nu/titanic1.jpg' height=50% width=50%>


The Titanic dataset is often used as an introduction to data analysis, especially for those interested in machine learning. Because we know who survived and who didn't, we can use the data on passengers to explore, look for trends that affect survival rate, and maybe even make some predictions on which passengers survived by looking at their data.   Today, we'll explore the Titanic dataset and get some real-world practice with cleaning and manipulating data.  

#### Pandas--Favorite Tool of Data Scientists 

<img src='https://media.giphy.com/media/aUhEBE0T8XNHa/giphy.gif' height=25% width=25%>

For data processing in Python, you can't beat the `pandas` library.  Pandas is used for creating dataframes, which are reminiscent of tables or spreadsheets, but much more powerful and easy to use.  Pandas provides a clean, easy-to-use API for reading, manipulating, sorting, and slicing data.  

#### Learning Goals

In this exercise, we have the following goals:

1.  Use pandas to read-in and manipulate data.
1.  Explore strategies for detecting outliers and dealing with missing (NaN) values
1.  Answer questions about our data set using descriptive statistics.
1.  Use pandas to slice our dataframe into smaller dataframes based on conditional logic (for instance, all female survivors under a certain age)

**Let's get started!**

(To execute any cell, press SHIFT + ENTER or press the 'Run' button on the toolbar at the top.)

We'll start by importing pandas and setting an alias:

In [2]:
# We'll be using pandas a lot, so let's alias it to the name 'pd' to save ourselves some keystrokes
import pandas as pd

Next, we'll want to read in our dataset from the from the the 'datasets' folder in this directory.  Did you know jupyter notebooks can also use terminal commands in code cells?  Just put a '%' sign before the command and it works the same as in the terminal.  For instance, if you wanted to list all the items contained in this current directory, you would just type `%ls`!  Try it below:

In [35]:
# Type % ls and run this cell
%

ERROR:root:Line magic function `%git` not found.


Our data set is contained within the `datasets` folder, and is called `titanic.csv`.  You may have some experience reading data in manually using the `open` keyword in python.  With large/complex datasets, this method can get tedious very quickly.  However, `pandas` makes this a simple task!

**TASK: Read in the titanic.csv file using the pd.read_csv( ) method in pandas.**

HINT:  You'll still need to pass the method the correct arguments--in order to get this right, You'll need the documentation for this method.  You can find that [here](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html).  

Alternatively, you can just type the method name with a question mark instead of parentheses to open a method's docs right here in the notebook!

In [5]:
# Need the docs? try typing pd.read_csv? 

path = "datasets/titanic.csv"
df =  pd.read_csv(path)  # store the newly created dataframe in this variable
df   # this will display the contents of df (the dataframe containing all our data, truncated for readability)

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
5              6         0       3   
6              7         0       1   
7              8         0       3   
8              9         1       3   
9             10         1       2   
10            11         1       3   
11            12         1       1   
12            13         0       3   
13            14         0       3   
14            15         0       3   
15            16         1       2   
16            17         0       3   
17            18         1       2   
18            19         0       3   
19            20         1       3   
20            21         0       2   
21            22         1       2   
22            23         1       3   
23            24         1       1   
24            25         0       3   
25            26         1       3   
26            27         0       3   
27            28         0       1   
28            29         1       3   
29            30         0       3   
..           ...       ...     ...   
861          862         0       2   
862          863         1       1   
863          864         0       3   
864          865         0       2   
865          866         1       2   
866          867         1       2   
867          868         0       1   
868          869         0       3   
869          870         1       3   
870          871         0       3   
871          872         1       1   
872          873         0       1   
873          874         0       3   
874          875         1       2   
875          876         1       3   
876          877         0       3   
877          878         0       3   
878          879         0       3   
879          880         1       1   
880          881         1       2   
881          882         0       3   
882          883         0       3   
883          884         0       2   
884          885         0       3   
885          886         0       3   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
5                                     Moran, Mr. James    male   NaN      0   
6                              McCarthy, Mr. Timothy J    male  54.0      0   
7                       Palsson, Master. Gosta Leonard    male   2.0      3   
8    Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                  Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
10                     Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
12                      Saundercock, Mr. William Henry    male  20.0      0   
13                         Andersson, Mr. Anders Johan    male  39.0      1   
14                Vestrom, Miss. Hulda Amanda Adolfina  female  14.0      0   
15                    Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   
16                                Rice, Master. Eugene    male   2.0      4   
17                        Williams, Mr. Charles Eugene    male   NaN      0   
18   Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.0      1   
19                             Masselmani, Mrs. Fatima  female   NaN

Great! With the pandas library, importing large datasets becomes really, really easy.  We can also use it to examine the data, and manipulate the dataset very easily. Before we can make sense of the data, we should probably have a feel for what each column means.  Here's the **Data Dictionary** explaining what each column and value actually means:


<center><h2>Data Dictionary</h2></center>


| Variable | Definition                                        | Key                                            |
|----------|---------------------------------------------------|------------------------------------------------|
| Survived | Survival                                          | 0 = No, 1 = Yes                                |
| Pclass   | Ticket Class (proxy for socio-economic status)    | 1 = 1st, 2 = 2nd, 3 = 3rd                      |
| Sex      | Sex                                               |                                                |
| Age      | Age (in years)                                    |                                                |
| SibSp    | # of siblings and/or spouses also aboard          |                                                |
| Parch    | # of parents / children also aboard               |                                                |
| Ticket   | Ticket number                                     |                                                |
| Fare     | Passenger fare (how much their ticket cost)       |                                                |
| Cabin    | Cabin number                                      |                                                |
| Embarked | Port of Embarkation (where they boarded the ship) | C = Cherbourg, Q = Queenstown, S = Southampton |

Let's get the basic descriptive statistics to see what we can figure out about this dataset.  

**TASK: Run the dataframe's .describe( ) command.  **

In [6]:
# Run df.describe().  What does each cell in the table mean?
df.describe()


PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

We can also call `.describe()` on individual columns (a single column of data is called a pandas _series_).  For instance, if we wanted to see the summary statistics on the _SibSp_ column, we would type `df["SibSp"].describe()`

See if you can answer the following questions:  

1.  How old is the oldest passenger on the titanic?
1.  How young is the youngest passenger?
1.  What is the average price paid for a ticket?
1.  How much did the most expensive ticket cost?
1.  How many passengers in the dataset are female?

Answers:

1.  80
1.  0.42
1.  32.20
1.  512.32
1.  889

In [34]:
# Use the .describe() method on the appropriate columns to answer the questions listed above.
df[(df["Sex"] == "female")].describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   314.000000  314.000000  314.000000  261.000000  314.000000   
mean    431.028662    0.742038    2.159236   27.915709    0.694268   
std     256.846324    0.438211    0.857290   14.110146    1.156520   
min       2.000000    0.000000    1.000000    0.750000    0.000000   
25%     231.750000    0.000000    1.000000   18.000000    0.000000   
50%     414.500000    1.000000    2.000000   27.000000    0.000000   
75%     641.250000    1.000000    3.000000   37.000000    1.000000   
max     889.000000    1.000000    3.000000   63.000000    8.000000   

            Parch        Fare  
count  314.000000  314.000000  
mean     0.649682   44.479818  
std      1.022846   57.997698  
min      0.000000    6.750000  
25%      0.000000   12.071875  
50%      0.000000   23.000000  
75%      1.000000   55.000000  
max      6.000000  512.329200

In [ ]:
# Use your knowledge of conditional slicing to answer the questions above!


<center><h3>Slicing</h3></center>

We can also use pandas to 'slice' dataframes, just as we would with a list in python.  However, unlike list slicing in vanilla python, we can slice using conditional logic.  For instance, what if we want to examine a dataframe that only contains the passengers that survived?  Easy-- we just type:

<center>`survived_df = df[df["Survived"] == 1]`</center>

The syntax for slicing can feel a bit clunky at first, but it will become intuitive with practice.  

If you want to slice on multiple conditions, you can do that too!  Just make sure each condition is wrapped in a set of parentheses.  For instance, if we wanted to grab a dataframe filled only when men that survived, we would use:
<br>
<br>
<center>`df[(df['Survived'] == 1) & (df['Sex'] == 'male')]`</center>


**TASK: Use your knowledge of conditional slicing to answer the following questions:**

(HINT: Don't forget about the `.describe()` method!)

1.  How many men survived?
1.  What is the average age of male passengers that survived?
1.  How many female passengers under the age of 30 did not survive?
1.  What was the most expensive ticket bought by answer passenger from Cherbourg that did NOT survive?
1.  Of all surviving passengers from Southampton, how many passengers paid between \$216 and \$676.50 for their ticket? 

**Answers:**

1.  109
1.  27.27
1.  42
1.  247.52
1.  93

In [56]:
df[(df['Embarked'] == 'C') & (df['Survived'] == 0) & (df['Fare'] > 200)]

PassengerId  Survived  Pclass                       Name   Sex   Age  \
118          119         0       1   Baxter, Mr. Quigg Edmond  male  24.0   
377          378         0       1  Widener, Mr. Harry Elkins  male  27.0   
557          558         0       1        Robbins, Mr. Victor  male   NaN   

     SibSp  Parch    Ticket      Fare    Cabin Embarked  
118      0      1  PC 17558  247.5208  B58 B60        C  
377      0      2    113503  211.5000      C82        C  
557      0      0  PC 17757  227.5250      NaN        C